File: PE-common/src/nl/adaptivity/util/activation/Sources.java
package nl.adaptivity.util.activation;

import net.devrieze.util.Streams;

import java.io.*;

import javax.xml.bind.util.JAXBSource;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Result;
import javax.xml.transform.Source;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.TransformerFactoryConfigurationError;
import javax.xml.transform.sax.SAXSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;


public final class Sources {

  private Sources() {}

  public static void writeToStream(final Source pSource, final OutputStream pOutputStream) throws TransformerException {
    writeToResult(pSource, new StreamResult(pOutputStream), false);
  }

  public static void writeToStream(final Source pSource, final OutputStream pOutputStream, boolean pIndent) throws TransformerException {
    writeToResult(pSource, new StreamResult(pOutputStream), pIndent);
  }

  public static void writeToWriter(final Source pSource, final Writer pWriter) throws TransformerException {
    writeToWriter(pSource, pWriter, false);
  }

  public static void writeToWriter(final Source pSource, final Writer pWriter, final boolean pIndent) throws TransformerException {
    writeToResult(pSource, new StreamResult(pWriter), pIndent);
  }

  public static void writeToResult(final Source pSource, final Result pResult) throws TransformerException{
    writeToResult(pSource, pResult, false);
  }

  public static void writeToResult(final Source pSource, final Result pResult, final boolean pIndent)
      throws TransformerFactoryConfigurationError, TransformerConfigurationException, TransformerException {
    final TransformerFactory factory = TransformerFactory.newInstance();
    final Transformer identityTransformer = factory.newTransformer();
    if (pIndent) {
      identityTransformer.setOutputProperty(OutputKeys.INDENT, "yes");
      identityTransformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
    }
    identityTransformer.transform(pSource, pResult);
  }

  public static InputStream toInputStream(final Source pSource) {
    if (pSource instanceof StreamSource) {
      final InputStream result = ((StreamSource) pSource).getInputStream();
      if (result!=null) { return result; }
    }
    if (pSource instanceof SAXSource && (! (pSource instanceof JAXBSource))) {
      final InputStream result = ((SAXSource) pSource).getInputSource().getByteStream();
      if (result!=null) { return result; }
    }
    final ByteArrayOutputStream baos = new ByteArrayOutputStream();
    try {
      Sources.writeToStream(pSource, baos);
    } catch (final TransformerException e) {
      throw new RuntimeException(e);
    }
    return new ByteArrayInputStream(baos.toByteArray());

  }

  public static Reader toReader(final Source pSource) {
    {
      if (pSource instanceof StreamSource) {
        Reader result = ((StreamSource) pSource).getReader();
        if (result!=null) { return result; }
      }
      if (pSource instanceof SAXSource && (! (pSource instanceof JAXBSource))) {
        final InputStream byteStream = ((SAXSource) pSource).getInputSource().getByteStream();
        if (byteStream!=null) { return new InputStreamReader(byteStream); }
      }
    }
    final CharArrayWriter caw = new CharArrayWriter();
    try {
      Sources.writeToWriter(pSource, caw);
    } catch (final TransformerException e) {
      throw new RuntimeException(e);
    }
    return new CharArrayReader(caw.toCharArray());
  }

  public static String toString(final Source pSource) {
    Reader in = null;
    try {
      if (pSource instanceof StreamSource) {
        Reader result = ((StreamSource) pSource).getReader();
        if (result!=null) { return Streams.toString(result); }
      }
      if (pSource instanceof SAXSource && (! (pSource instanceof JAXBSource))) {
        final InputStream byteStream = ((SAXSource) pSource).getInputSource().getByteStream();
        if (byteStream!=null) { return Streams.toString(new InputStreamReader(byteStream)); }
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    final StringWriter sw = new StringWriter();
    try {
      Sources.writeToWriter(pSource, sw);
    } catch (final TransformerException e) {
      throw new RuntimeException(e);
    }
    return sw.toString();
  }
}

File: PE-common/src/jvmMain/kotlin/nl/adaptivity/util/activation/Sources.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.util.activation


import net.devrieze.util.readString
import java.io.*
import javax.xml.bind.util.JAXBSource
import javax.xml.transform.*
import javax.xml.transform.sax.SAXSource
import javax.xml.transform.stream.StreamResult
import javax.xml.transform.stream.StreamSource

@JvmOverloads
fun Source.writeToStream(outputStream: OutputStream, indent: Boolean = false) {
    this.writeToResult(StreamResult(outputStream), indent)
}

@JvmOverloads
fun Source.writeToWriter(writer: Writer, indent: Boolean = false) {
    this.writeToResult(StreamResult(writer), indent)
}

@JvmOverloads
fun Source.writeToResult(result: Result, indent: Boolean = false) {
    val factory = TransformerFactory.newInstance()
    val identityTransformer = factory.newTransformer()
    if (indent) {
        identityTransformer.setOutputProperty(OutputKeys.INDENT, "yes")
        identityTransformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2")
    }
    identityTransformer.transform(this, result)
}

fun Source.toInputStream(): InputStream {
    if (this is StreamSource) {
        inputStream?.let { return it }
    }
    if (this is SAXSource && this !is JAXBSource) {
        inputSource.byteStream?.let { return it }
    }

    val byteArray = ByteArrayOutputStream().also { baos ->
        writeToStream(baos)
    }.toByteArray()
    return ByteArrayInputStream(byteArray)
}

fun Source.toReader(): Reader {
    if (this is StreamSource) {
        reader?.let { return it }
    }
    if (this is SAXSource && this !is JAXBSource) {
        inputSource.byteStream?.let { return InputStreamReader(it) }
    }

    val charArray = CharArrayWriter().also { caw -> writeToWriter(caw) }.toCharArray()
    return CharArrayReader(charArray)

}

@JvmOverloads
fun Source.toString(indent: Boolean = false): String {

    if (this is StreamSource) {
        reader?.let { return it.readString() }
    }
    if (this is SAXSource && this !is JAXBSource) {
        inputSource.byteStream?.let { return InputStreamReader(it).readString() }
    }

    return StringWriter().also { writer ->
        writeToWriter(writer, indent)
    }.toString()
}

object Sources {

    @Deprecated("Use extension function", ReplaceWith("source.writeToStream(outputStream)"))
    fun writeToStream(source: Source, outputStream: OutputStream) {
        source.writeToStream(outputStream)
    }

    @Deprecated("Use extension function", ReplaceWith("source.writeToStream(outputStream, indent)"))
    fun writeToStream(source: Source, outputStream: OutputStream, indent: Boolean) {
        source.writeToStream(outputStream, indent)
    }

    @Deprecated("Use extension function", ReplaceWith("source.writeToWriter(writer, indent)"))
    @JvmOverloads
    fun writeToWriter(source: Source, writer: Writer, indent: Boolean = false) {
        source.writeToWriter(writer, indent)
    }

    @Deprecated("Use extension function", ReplaceWith("source.writeToResult(result, indent)"))
    @JvmOverloads
    fun writeToResult(source: Source, result: Result, indent: Boolean = false) {
        source.writeToResult(result, indent)
    }

    @Deprecated("Use extension function", ReplaceWith("source.toInputStream()"))
    fun toInputStream(source: Source): InputStream {
        return source.toInputStream()
    }

    @Deprecated("Use extension function", ReplaceWith("source.toReader()"))
    fun toReader(source: Source): Reader {
        return source.toReader()
    }

    @Deprecated("Use extension function", ReplaceWith("source.toString(indent)"))
    @JvmOverloads
    fun toString(source: Source, indent: Boolean = false): String {
        return source.toString(indent)
    }
}
